homeassistant/loader.py

```diff
@@ -908,75 +917,123 @@ def get_component(self) -> ComponentProtocol:
+        platforms = await self.async_get_platforms([platform_name])
+        return platforms[platform_name]
+
+    async def async_get_platforms(
+        self, platform_names: Iterable[Platform | str]
+    ) -> dict[str, ModuleType]:
+        """Return a platforms for an integration."""
         domain = self.domain
-        full_name = f"{self.domain}.{platform_name}"
-        if platform := self._get_platform_cached(full_name):
-            return platform
-        if future := self._import_futures.get(full_name):
-            return await future
-        if debug := _LOGGER.isEnabledFor(logging.DEBUG):
-            start = time.perf_counter()
-        import_future = self.hass.loop.create_future()
-        self._import_futures[full_name] = import_future
-        load_executor = (
-            self.import_executor
-            and domain not in self.hass.config.components
-            and f"{self.pkg_path}.{domain}" not in sys.modules
-        )
-        try:
-            if load_executor:
-                try:
-                    platform = await self.hass.async_add_import_executor_job(
-                        self._load_platform, platform_name
-                    )
-                except ImportError as ex:
+        platforms: dict[str, ModuleType] = {}
+
+        load_executor_platforms: list[str] = []
+        load_event_loop_platforms: list[str] = []
+        in_progress_imports: dict[str, asyncio.Future[ModuleType]] = {}
+        import_futures: list[tuple[str, asyncio.Future[ModuleType]]] = []
+
+        for _plat in platform_names:
+            platform_name = _plat.value if type(_plat) is Platform else _plat
+            full_name = f"{domain}.{platform_name}"
+            if platform := self._get_platform_cached(full_name):
+                platforms[platform_name] = platform
+                continue
+
+            # Another call to async_get_platforms is already importing this platform
+            if future := self._import_futures.get(platform_name):
+                in_progress_imports[platform_name] = future
+                continue
+
+            if (
+                self.import_executor
+                and full_name not in self.hass.config.components
+                and f"{self.pkg_path}.{platform_name}" not in sys.modules
+            ):
+                load_executor_platforms.append(platform_name)
+            else:
+                load_event_loop_platforms.append(platform_name)
+
+            import_future = self.hass.loop.create_future()
+            self._import_futures[platform_name] = import_future
+            import_futures.append((platform_name, import_future))
+
+        if load_executor_platforms or load_event_loop_platforms:
+            if debug := _LOGGER.isEnabledFor(logging.DEBUG):
+                start = time.perf_counter()
+
+            try:
+                if load_executor_platforms:
+                    try:
+                        platforms.update(
```

Comment:
Do we need to handle a single platform in a group failing to load? Or is it ok if we blow up in any of them fail to load?

Reply:
I think it might be best to fail rather than try to handle partial success as we know things are pretty broken in that case and at that point it's debatable how useful a semi broken integration is

Reply:
Agreed, blow up hard.